### PR TITLE
Fix query failed with session pool when reconnected

### PR DESCRIPTION
**Pull request by songqing**
## What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### Description:
When querying with session pool, the query can use retryConnect() when the connection is broken.
However, after reconnected, the client only send "USE space" statement without the original query, so, the original query will be failed as it is not executed.

## How do you solve it?
After reconnected, the client will send a query not only with "USE space", but also with the original statment.

### CODE REVIEW

**Comment by songqing** [comment]
@Shylock-Hg Could you help to review this PR?

File: src/client/SessionPool.cpp [hunk 1]
```diff
@@ -48,8 +49,9 @@ ExecutionResponse SessionPool::executeWithParameter(
   if (result.second) {
     auto resp = result.first.executeWithParameter(stmt, parameters);
     if (resp.spaceName != nullptr && *resp.spaceName != config_.spaceName_) {
-      // switch to origin space
-      result.first.execute("USE " + config_.spaceName_);
+      // switch to origin space and try to query again
```
Comment (Shylock-Hg) [inline src/client/SessionPool.cpp]: What's the error? It's by design, only switch to origin space when space changed.

Comment (songqing) [reply src/client/SessionPool.cpp]: This error will be happened, "Space was not chosen.". 
Just as what I said above, user's query will be failed as only do a query "USE space", but the original stmt is not executed.

Comment (Shylock-Hg) [reply src/client/SessionPool.cpp]: I can't get your point, please list your operations step by step.

Comment (songqing) [reply src/client/SessionPool.cpp]: 1 When a query failed as the connection is broken, we can use retryConnect() to establish a new connection now, but the query will still fail as "Space was not chosen".  if we only retry here with "USE space", the original stmt will fail as it's not run.
2 "if (resp.spaceName != nullptr && *resp.spaceName != config_.spaceName_) ", this statement will only happen when connection is broken and reconnect, as session pool is only inited with ONE config space name. space name will not changed after inited, see [here](https://github.com/vesoft-inc/nebula-python/blob/1fe9a44aa7b0f9fdfc35fec12af63e30d7f3546e/README.md?plain=1#L102)
"Each session pool is corresponding to a single USER and a single Space"

Comment (Shylock-Hg) [reply src/client/SessionPool.cpp]: Why report "Space was not chosen", the `use space` is requested?

Comment (songqing) [reply src/client/SessionPool.cpp]: Yes, it seems almost all the nGQL need "use space" first(only a small number of statements don't, such as "yield 1"), in our case, we use "GO" statement to query.

Comment (Shylock-Hg) [reply src/client/SessionPool.cpp]: Yes, I mean `SessionPool` will switch space automatically.

Comment (Shylock-Hg) [reply src/client/SessionPool.cpp]: This `if block` is executed when query return `ok` and space changed. So I don't know why it's retried?

Comment (songqing) [reply src/client/SessionPool.cpp]: Maybe you still don't get my idea, I mean "Each session pool is corresponding to a single USER and a single Space", 
https://github.com/vesoft-inc/nebula-python/blob/1fe9a44aa7b0f9fdfc35fec12af63e30d7f3546e/README.md?plain=1#L102

Comment (songqing) [reply src/client/SessionPool.cpp]: I mean with session pool, there is no "space changed" operation

Comment (Shylock-Hg) [reply src/client/SessionPool.cpp]: I know this.

Comment (Shylock-Hg) [reply src/client/SessionPool.cpp]: > I mean with session pool, there is no "space changed" operation

There is no such limitation in the program, user can run any query include `use space`. It's just a usage limitation.

Comment (songqing) [reply src/client/SessionPool.cpp]: > This `if block` is executed when query return `ok` and space changed. So I don't know why it's retried?

The query return is not ok, it's an error "Space was not choosen"

Comment (songqing) [reply src/client/SessionPool.cpp]: > > This `if block` is executed when query return `ok` and space changed. So I don't know why it's retried?
> 
> The query return is not ok, it's an error "Space was not choosen"

Only return the error  "Space was not choosen", then, the if block is executed

Comment (Shylock-Hg) [reply src/client/SessionPool.cpp]: What's your query?

Comment (songqing) [reply src/client/SessionPool.cpp]: > > I mean with session pool, there is no "space changed" operation
> 
> There is no such limitation in the program, user can run any query include `use space`. It's just a usage limitation.
In the link above, there is a description，"If you need to run queries in different spaces, you may have multiple session pools.", so, maybe users will use like that.
Besides, this PR fix a bug as I described before, when connection is broken, even if reconnected(), the original query will fail with the error "Space was not chosen".

Comment (songqing) [reply src/client/SessionPool.cpp]: > What's your query?

My query is "go from xx yield xx" and without using "Use space" every time

**Comment by Shylock-Hg** [comment]
You should switch to origin space after `retryConnect` in `SessionPool`

**Comment by songqing** [comment]
> You should switch to origin space after `retryConnect` in `SessionPool`
OK, but that will cause a query error.
retryConnect() is executed automatically, users couldn't feel it. When users get an error "Space was not chosen", then the query was failed, we don't want to let the error occur.
And with this PR, users' query will not return an error.

**Comment by Shylock-Hg** [comment]
> > You should switch to origin space after `retryConnect` in `SessionPool`
> > OK, but that will cause a query error.
> > retryConnect() is executed automatically, users couldn't feel it. When users get an error "Space was not chosen", then the query was failed, we don't want to let the error occur.
> > And with this PR, users' query will not return an error.

  I mean you should add this logical in this PR, instead of current modification. Current code block is not to do this.

**Comment by Shylock-Hg** [comment]
Finally, all sessions in `SessionPool` is bind to an user and one space when user request a query. The session constructed by `retryConnect` doesn't bind space, so leads to this error. Just fix the root error.

**Comment by songqing** [comment]
> 



> > > You should switch to origin space after `retryConnect` in `SessionPool`
> > > OK, but that will cause a query error.
> > > retryConnect() is executed automatically, users couldn't feel it. When users get an error "Space was not chosen", then the query was failed, we don't want to let the error occur.
> > > And with this PR, users' query will not return an error.
> 
> I mean you should add this logical in this PR, instead of current modification. Current code block is not to do this.

In fact, I tried to do this, but it's difficult as the Session class has no member "SpaceName"

**Comment by Shylock-Hg** [comment]
> > 
> 
> > > > You should switch to origin space after `retryConnect` in `SessionPool`
> > > > OK, but that will cause a query error.
> > > > retryConnect() is executed automatically, users couldn't feel it. When users get an error "Space was not chosen", then the query was failed, we don't want to let the error occur.
> > > > And with this PR, users' query will not return an error.
> > 
> > 
> > I mean you should add this logical in this PR, instead of current modification. Current code block is not to do this.
> 
> In fact, I tried to do this, but it's difficult as the Session class has no member "SpaceName"

I think you could add a `CallBack` of reconnect parameter to `Session::execute`

**Comment by songqing** [comment]
> Finally, all sessions in `SessionPool` is bind to an user and one space when user request a query. The session constructed by `retryConnect` doesn't bind space, so leads to this error. Just fix the root error.

retryConnect() is executed in Session.cc, there is no member "SpaceName" in this class, if we add a such member, much code will be changed.
Besides, Session.cc don't need a member "SpaceName" as some query like "yield 1" don't need the space name, only session pool need a "SpaceName"

**Comment by songqing** [comment]
> 



> > > 
> > 
> > 
> > > > > You should switch to origin space after `retryConnect` in `SessionPool`
> > > > > OK, but that will cause a query error.
> > > > > retryConnect() is executed automatically, users couldn't feel it. When users get an error "Space was not chosen", then the query was failed, we don't want to let the error occur.
> > > > > And with this PR, users' query will not return an error.
> > > 
> > > 
> > > I mean you should add this logical in this PR, instead of current modification. Current code block is not to do this.
> > 
> > 
> > In fact, I tried to do this, but it's difficult as the Session class has no member "SpaceName"
> 
> I think you could add a `CallBack` of reconnect parameter to `Session::execute`
"If you need to run queries in different spaces, you may have multiple session pools.", if this works, I think this PR can fix the bug.
Or, there maybe too much change should do

**Comment by songqing** [comment]
@Shylock-Hg Could you help review again? The original logic is reserved, and adding a new condition to fix the problem

**Review by Shylock-Hg — APPROVED** [review]

**Pull request merged**